src/core/Akka.Streams/Dsl/RestartSource.cs

```diff
@@ -0,0 +1,188 @@
+        /// <para>
+        /// This <see cref="Source"/> will never emit a failure, since the failure of the wrapped <see cref="Source"/> is always handled by
+        /// restarting. The wrapped <see cref="Source"/> can be cancelled by cancelling this <see cref="Source"/>.
+        /// When that happens, the wrapped <see cref="Source"/>, if currently running will be cancelled, and it will not be restarted.
+        /// This can be triggered simply by the downstream cancelling, or externally by introducing a <see cref="IKillSwitch"/> right
+        /// after this <see cref="Source"/> in the graph.
+        /// </para>
+        /// <para>This uses the same exponential backoff algorithm as <see cref="BackoffOptions"/>.</para>
+        /// </summary>
+        /// <param name="sourceFactory">A factory for producing the <see cref="Source"/> to wrap.</param>
+        /// <param name="settings"><see cref="RestartSettings" /> defining restart configuration</param>
+        public static Source<T, NotUsed> OnFailuresWithBackoff<T, TMat>(Func<Source<T, TMat>> sourceFactory, RestartSettings settings)
+            => Source.FromGraph(new RestartWithBackoffSource<T, TMat>(sourceFactory, settings, onlyOnFailures: true));
+    }
+    
+    internal sealed class RestartWithBackoffSource<T, TMat> : GraphStage<SourceShape<T>>
+    {
+        public Func<Source<T, TMat>> SourceFactory { get; }
+        public RestartSettings Settings { get; }
+        public bool OnlyOnFailures { get; }
+
+        public RestartWithBackoffSource(Func<Source<T, TMat>> sourceFactory, RestartSettings settings, bool onlyOnFailures)
+        {
+            SourceFactory = sourceFactory;
+            Settings = settings;
+            OnlyOnFailures = onlyOnFailures;
+            Shape = new SourceShape<T>(Out);
+        }
+
+        public Outlet<T> Out { get; } = new Outlet<T>("RestartWithBackoffSource.out");
+
+        public override SourceShape<T> Shape { get; }
+
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes) => new Logic(this, inheritedAttributes, "Source");
+
+        private sealed class Logic : RestartWithBackoffLogic<SourceShape<T>, T, T>
+        {
+            private readonly RestartWithBackoffSource<T, TMat> _stage;
+            private readonly Attributes _inheritedAttributes;
+
+            public Logic(RestartWithBackoffSource<T, TMat> stage, Attributes inheritedAttributes, string name)
+                : base(name, stage.Shape, null, stage.Out, stage.Settings, stage.OnlyOnFailures)
+            {
+                _stage = stage;
+                _inheritedAttributes = inheritedAttributes;
+                Backoff();
+            }
+
+            protected override void StartGraph()
+            {
+                var sinkIn = CreateSubInlet(_stage.Out);
+                SubFusingMaterializer.Materialize(_stage.SourceFactory().To(sinkIn.Sink), _inheritedAttributes);
```

Comment:
Brought this in line with the scala sourcecode. IMO this used to contain a big ass bug. Before it never used the SourceFactory. How that ever worked i don't know. I can only guess the specs surrounding the RestartSource are bad.

Reply:
Given how racy those specs have been, I think that's a good assumption